### DISCUSSION TITLE
how to use informer to list k8s object,like to list pods?

### DISCUSSION THREAD

**Discussion by vCassius** [kube-rs/kube#1282] [General]
apiserver would be pressure if all requests went to it,so i want to use informer to cache it.

**Comment by clux** [comment]
the informer equivalent is to pass a watcher stream through a reflector and giving it the write access to a store.
see e.g. node_reflector example: https://github.com/kube-rs/kube/blob/5e98a92c881b9387af9e0f32cc38b92246a5aa7e/examples/node_reflector.rs#L20-L41

**Reply by vCassius** [reply]
thx! i will try to this.

**Reply by vCassius** [reply]
i'm tried it succeed,but anohter question. when i stored event and pod, it's will not auto clean when the pod or event object destroy by kubernetes. then,my server memory used will be grow up...

**Reply by clux** [reply]
i'm not sure if you get removed events from watching `Event` actually.. that resource is generally considered append-only and then kubernetes wipes them without ceremony once they reach a certain age (like 1h).

it is possible you have to store events in a different way - e.g. forwarding all watcher objects to a vec - and periodically prune elements with a creationTimestamp > some_threshold

**Reply by vCassius** [reply]
hey guy, is there any way to clone store writer? ,cuz i want to clean events from store,but writer can nit move. and no way to clone it.

**Reply by clux** [reply]
the writer is deliberately not clone-able, see [doc str for it](https://docs.rs/kube-runtime/0.85.0/src/kube_runtime/reflector/store.rs.html#15-18).

if you are customizing how storage and cleanup is done, you probably want to implement you own "writer-like" thing, with your own "apply-fn": https://docs.rs/kube-runtime/0.85.0/src/kube_runtime/reflector/store.rs.html#61-84 and call that on every event [like reflector does](https://docs.rs/kube-runtime/0.85.0/src/kube_runtime/reflector/mod.rs.html#98)

**Reply by vCassius** [reply]
thx, i will try it with apply_watcher_event

**Reply by vCassius** [reply]
seems event informer working,but i got this err:
  2023-09-11 11:57:33.671 ERROR v_kubernetes_action::pkg::global::objects: event try next exception: error returned by apiserver during watch: The resourceVersion for the provided watch is too old.: Expired
    at src/pkg/global/objects.rs:95

**Reply by clux** [reply]
i'm guessing if you don't have enough events flowing through the system, the normal watcher attempt to resume from a (now wiped) resume point will not work. because events wipe so frequently, the watcher will probably have to restart from "0" again a lot more frequently than a watcher for any other resource.

**Reply by clux** [reply]
so if you see every event being passed through then it __shouldn't__ be anything to worry about in your case - you could filter out the error by matching on it if it's annoying.

**Comment by vCassius** [comment]
it's a way to forwarding all watcher objects to a vec then clean, when i used client-go to create informer, It automatically manages objects when the kubernetes create or destory object.
